src/qt3utils/datagenerators/princeton.py

```diff
@@ -0,0 +1,339 @@
+
+        else:
+
+            # TODO: might need to add a proper exception
+            print('Invalid setting:{}'.format(setting))
+
+        return
+
+    def get(self, setting):
+        """
+        Helper function for getting experiment parameters with a basic check
+        to see if a specific setting is valid.
+        """
+
+        if self.experiment.Exists(setting):
+
+            value = self.experiment.GetValue(setting)
+
+        else:
+
+            value = []
+            print('Invalid setting: {}'.format(setting))
+
+        return value
+
+    def load_experiment(self, value):
+        self.experiment.Load(value)
+
+    def acquire(self):
+        """
+        Helper function to acquire the data from the PIXIS-style cameras.
+        Need to check array reorganization since there is only a single vertical pixel.
+        For a single frame, the data is returned as a 2D array (just raw counts
+        from each pixel). Taking a section in the vertical (400-pixel) direction
+        corresponds to wavelength averaging.
+        For multiple frames, each exposure is stacked in a third dimension, so
+        averaging can be performed simply by summing over the different planes.
+        """
+
+        acquisition_time = self.get(lf.AddIns.CameraSettings.ShutterTimingExposureTime) #to allow the camera to capture the desired amount of data during the specified exposure time.
+        num_frames = self.get(lf.AddIns.ExperimentSettings.FrameSettingsFramesToStore)
+
+        self.experiment.Acquire()
+
+        sleep(0.001 * acquisition_time * num_frames)  #sleep delay that waits for the exposure duration of the camera
+
+        while self.experiment.IsRunning:
+            sleep(0.1)  #loop that repeatedly checks whether the experiment is still running so it decided when to move to next block of code
+
+        last_file = self.application.FileManager.GetRecentlyAcquiredFileNames().get_Item(0)
+
+        frame_count = self.application.FileManager.OpenFile(last_file, FileAccess.Read)
+
+        if frame_count.Regions.Length == 1:
+
+            if frame_count.Frames == 1:
+
+                frame = frame_count.GetFrame(0, 0)
+                data = np.reshape(np.fromiter(frame.GetData(), 'uint16'), [frame.Width, frame.Height], order='F')
+
+            else:
+                data = np.array([])
+                for i in range(0, frame_count.Frames):
+                    frame = frame_count.GetFrame(0, i)
+                    new_frame = np.fromiter(frame.GetData(), 'uint16')
+
+                    new_frame = np.reshape(np.fromiter(frame.GetData(), 'uint16'), [frame.Width, frame.Height],
+                                           order='F')
+                    data = np.dstack((data, new_frame)) if data.size else new_frame
+
+            return data
+
+
+        else:
+            
+            print('frame_count.Regions is not. Please retry.')
+            print(frame_count.Frames)
+        
```

Comment:
This function ends without returning anything. Python is super flexible, which is great in some ways and problematic in others. For this function, the caller is expecting to get a 2D numpy array. I think you should return one, no matter what, even if it's empty.   If there's a serious error, then you can raise an Exception to force the caller to handle it. (And you should probably document